### PR TITLE
Refine endpoints in system-api plugin [ECR-4154]

### PR DESCRIPTION
**Pull request by aleksuss**
## Overview

---

See: https://jira.bf.local/browse/ECR-4154

### Definition of Done

- [ ] There are no TODOs left in the merged code
- [ ] Change is covered by automated tests
- [ ] Benchmark results are attached (if applicable)
- [ ] The [coding guidelines] are followed
- [ ] Public API has proper documentation
- [ ] Changelog is updated if needed (in case of notable or breaking changes)
- [ ] The continuous integration build passes

[coding guidelines]: https://github.com/exonum/exonum/blob/master/CONTRIBUTING.md#conventions

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,21 @@ The project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)
...
+  been changed. (#1790) Current list of the endpoints:
+  - `v1/info` - obtains information about the node;
+  - `v1/stats` - obtains statistics of the node;
+  - `v1/peers` - adds a peer to the Exonum node;
+  - `v1/consensus_status` - enables or disables consensus on the node;
+  - `v1/shutdown` - shuts down the node.
+
+#### exonum-supervisor
+
+- `supervisor/services` has been added which obtains information about
```
Comment (slowli) [inline CHANGELOG.md]: Nit / grammar: `supervisor/services` **endpoint** has been added?

File: cli/tests/node.rs [hunk 2]
```diff
@@ -106,7 +106,7 @@ fn node_basic_workflow() -> Result<(), failure::Error> {
     let client = reqwest::Client::new();
     // Check info returned by the system API plugin.
```
Comment (slowli) [inline cli/tests/node.rs]: This comment should be updated.

File: components/system-api/src/private.rs [hunk 3]
```diff
@@ -10,143 +10,108 @@
 // distributed under the License is distributed on an "AS IS" BASIS,
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
-// limitations under the License.
```
Comment (slowli) [inline components/system-api/src/private.rs]: Why is this removed?

File: components/system-api/src/private.rs [hunk 4]
```diff
@@ -10,143 +10,108 @@
...
+//! # Get Node Statistics
 //!
 //! | Property    | Value |
 //! |-------------|-------|
-//! | Path        | `/api/system/v1/network` |
+//! | Path        | `/api/system/v1/stats` |
 //! | Method      | GET   |
 //! | Query type  | - |
-//! | Return type | [`NodeInfo`] |
+//! | Return type | [`StatsInfo`] |
```
Comment (slowli) [inline components/system-api/src/private.rs]: Nit: I'd call the type `NodeStats`.

File: components/system-api/src/private.rs [hunk 5]
```diff
@@ -206,59 +171,80 @@
...
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
-pub struct ServiceInfo {
-    /// Service name.
-    pub name: String,
-    /// Service identifier for the database schema and service messages.
-    pub id: InstanceId,
+/// Information about the current state of the node.
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
+pub struct StatsInfo {
+    /// Height of the last committed block.
```
Comment (slowli) [inline components/system-api/src/private.rs]: Nit: I'd rephrase as "Height of the blockchain."

File: components/system-api/src/private.rs [hunk 6]
```diff
@@ -206,59 +171,80 @@
...
+    pub height: u64,
+    /// Total number of uncommitted transactions stored in persistent pool.
+    pub tx_pool_size: u64,
+    /// Total number of transactions in the blockchain.
+    pub tx_count: u64,
+    /// Size of the transaction cache.
+    pub tx_cache_size: usize,
+    /// Work duration of the node.
```
Comment (slowli) [inline components/system-api/src/private.rs]: In what units? (seconds? millis?)

File: components/system-api/src/private.rs [hunk 7]
```diff
@@ -206,59 +171,80 @@
...
+    pub uptime: u64,
 }
 
-/// Short information about the current node.
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
-pub struct NodeInfo {
-    /// Version of the `exonum` crate.
-    pub core_version: Option<String>,
+/// Information about whether it is possible to achieve the consensus between
```
Comment (slowli) [inline components/system-api/src/private.rs]: Off-topic: Is this information garbage if the node is an auditor? Also, "Consensus enabled and the node has enough connected peers" does **not** mean that consensus can necessarily advance; for example, some of peers may have consensus disabled.

File: components/system-api/src/private.rs [hunk 8]
```diff
@@ -206,59 +171,80 @@
...
-    }
+/// Info about connected peer.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+pub struct ConnectedPeerInfo {
+    /// Address of the peer.
+    pub address: String,
+    /// Consensus public key of the peer.
+    pub public_key: PublicKey,
+    /// Connect direction.
+    pub direction: String,
```
Comment (slowli) [inline components/system-api/src/private.rs]: Why is this a string and not an enum with 2 variants?

File: components/system-api/src/private.rs [hunk 9]
```diff
@@ -206,59 +171,80 @@
...
 }
 
-/// Short information about the current node.
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
-pub struct NodeInfo {
-    /// Version of the `exonum` crate.
-    pub core_version: Option<String>,
+/// Information about whether it is possible to achieve the consensus between
+/// validators in the current state.
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq)]
```
Comment (slowli) [inline components/system-api/src/private.rs]: We customarily use `#[serde(rename_all = "snake_case")]` for enums.

File: components/system-api/src/private.rs [hunk 10]
```diff
@@ -206,59 +171,80 @@
...
+/// Short information about the current node.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+pub struct NodeInfo {
+    /// Consensus status.
+    pub consensus_status: ConsensusStatus,
+    /// List of connected peers.
+    pub connected_peers: Vec<ConnectedPeerInfo>,
+    /// Version of the `exonum` crate.
+    pub exonum_version: String,
+    /// Version of thr Rust.
```
Comment (slowli) [inline components/system-api/src/private.rs]: Grammar: "Version of thr Rust" -> "Rust version"

File: components/system-api/src/private.rs [hunk 11]
```diff
@@ -206,59 +171,80 @@
...
-    pub outgoing_connections: HashMap<SocketAddr, OutgoingConnection>,
+/// Short information about the current node.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+pub struct NodeInfo {
+    /// Consensus status.
+    pub consensus_status: ConsensusStatus,
+    /// List of connected peers.
+    pub connected_peers: Vec<ConnectedPeerInfo>,
+    /// Version of the `exonum` crate.
+    pub exonum_version: String,
```
Comment (slowli) [inline components/system-api/src/private.rs]: Could this be made into a `semver::Version`?

File: exonum/src/helpers/user_agent.rs [hunk 12]
```diff
@@ -32,10 +32,29 @@ pub fn user_agent() -> String {
...
+/// Returns OS info of host on which run the node.
+#[doc(hidden)]
+pub fn os_info() -> String {
+    os_info::get().to_string()
+}
+
+/// Returns a version of the exonum framework.
+#[doc(hidden)]
+pub fn exonum_version() -> Option<String> {
+    let versions: Vec<_> = USER_AGENT.split('/').collect();
```
Comment (slowli) [inline exonum/src/helpers/user_agent.rs]: Using `next()` instead of `collect()` would be more efficient.

File: exonum/src/helpers/user_agent.rs [hunk 13]
```diff
@@ -32,10 +32,29 @@ pub fn user_agent() -> String {
...
+    versions.get(0).map(|v| v.trim().to_owned())
+}
+
+/// Returns a version of the rust compiler.
+#[doc(hidden)]
+pub fn rust_version() -> Option<String> {
+    let versions: Vec<_> = USER_AGENT.split('/').collect();
```
Comment (slowli) [inline exonum/src/helpers/user_agent.rs]: Using `nth(1)` instead of `collect()` would be more efficient.

File: components/system-api/src/private.rs [hunk 14]
```diff
@@ -269,57 +255,85 @@ pub struct ConsensusEnabledQuery {
...
+                os_info: os_info(),
             })
         });
-        self
+        self_
     }
 
-    fn handle_peer_add(self, name: &'static str, api_scope: &mut ApiScope) -> Self {
+    fn handle_stats(self, name: &'static str, api_scope: &mut ApiScope) -> Self {
+        let self_ = self.clone();
```
Comment (slowli) [inline components/system-api/src/private.rs]: Micro-nit / bikeshedding: IMO, moving `self_` into closure would be more idiomatic (you could also call the variable `this` if the underscore at the end looks bad).

File: components/system-api/src/private.rs [hunk 15]
```diff
@@ -269,57 +255,85 @@ pub struct ConsensusEnabledQuery {
...
+                connected_peers.push(ConnectedPeerInfo::new(&connect_info, "incoming"));
             }
 
-            Ok(PeersInfo {
-                incoming_connections: shared_api_state.incoming_connections(),
-                outgoing_connections,
+            Ok(NodeInfo {
+                consensus_status: self.get_consensus_status(),
+                connected_peers,
+                exonum_version: exonum_version().unwrap_or_else(|| "unknown".to_owned()),
```
Comment (slowli) [inline components/system-api/src/private.rs]: I think this and the following vars should be instantiated once and then cloned.

File: components/system-api/src/private.rs [hunk 16]
```diff
@@ -269,57 +255,85 @@ pub struct ConsensusEnabledQuery {
...
+
+            for connect_info in shared_api_state.incoming_connections() {
+                connected_peers.push(ConnectedPeerInfo::new(&connect_info, "incoming"));
             }
 
-            Ok(PeersInfo {
-                incoming_connections: shared_api_state.incoming_connections(),
-                outgoing_connections,
+            Ok(NodeInfo {
+                consensus_status: self.get_consensus_status(),
```
Comment (slowli) [inline components/system-api/src/private.rs]: Nit: `get_consensus_status()` can be made a free function or a static method accepting `&SharedNodeState` as an argument. In this way, you wouldn't need to clone `self` above.

File: exonum/src/helpers/user_agent.rs [hunk 17]
```diff
@@ -48,4 +67,16 @@ mod tests {
             assert!(!val.is_empty());
         }
     }
+
+    #[test]
+    fn check_exonum_versions() {
+        let exonum_version = exonum_version().unwrap();
+        assert!(exonum_version.contains("exonum"));
```
Comment (slowli) [inline exonum/src/helpers/user_agent.rs]: What's the point of including "exonum" prefix into `exonum_version()`? Same question regarding the "rust" prefix for `rust_version`.

File: .travis.yml [hunk 18]
```diff
@@ -132,6 +132,10 @@ jobs:
   - name: unit-test
     script:
     - cargo test --all --lib --tests --benches
+
+  # Doc tests.
+  - name: doc-test
```
Comment (slowli) [inline .travis.yml]: Good idea, although inability to share build cache among this job and unit tests can increase build time for this job (but, hopefully, not the overall build time).

File: components/system-api/src/private.rs (hunk 7 above)
Comment (aleksuss) [reply components/system-api/src/private.rs]: Yeah. I'll rephrase this description.

File: components/system-api/src/private.rs (hunk 11 above)
Comment (aleksuss) [reply components/system-api/src/private.rs]: Potentially it could. But why ?

File: .travis.yml (hunk 18 above)
Comment (aleksuss) [reply .travis.yml]: Additionally, we use `sscache` for caching which stores its cache to `$HOME/.cache` and this directory shares between jobs. I suppose. By the way. I try to remove .cargo caching. I think in our case it works not optimal and consumes more time for archiving/unarchiving.

File: components/system-api/src/private.rs (hunk 15 above)
Comment (aleksuss) [reply components/system-api/src/private.rs]: But there is a case when we don't need this var at all. Redundant instantiation ?

File: components/system-api/src/private.rs [hunk 19]
```diff
@@ -19,134 +19,101 @@
...
+//! # Get Node Statistics
 //!
 //! | Property    | Value |
 //! |-------------|-------|
-//! | Path        | `/api/system/v1/network` |
+//! | Path        | `/api/system/v1/stats` |
 //! | Method      | GET   |
 //! | Query type  | - |
-//! | Return type | [`NodeInfo`] |
+//! | Return type | [`StatsInfo`] |
```
Comment (slowli) [inline components/system-api/src/private.rs]: Type name and link should be updated here to point to `NodeStats`. Weird that `deadlinks` didn't pick up this error. (`cargo clean --doc` should probably run before generating docs.)

File: exonum/build.rs [hunk 20]
```diff
@@ -42,10 +42,9 @@ fn create_path_to_protobuf_schema_env() {
 }
 
 fn write_user_agent_file() {
-    let package_name = option_env!("CARGO_PKG_NAME").unwrap_or("exonum");
-    let package_version = option_env!("CARGO_PKG_VERSION").unwrap_or("?");
+    let exonum_version = option_env!("CARGO_PKG_VERSION").unwrap_or("?");
     let rust_version = rust_version().unwrap_or_else(|| "rust ?".to_string());
```
Comment (slowli) [inline exonum/build.rs]: `rust ` prefix should be removed here.

File: exonum/build.rs [hunk 21]
```diff
@@ -76,5 +75,7 @@ fn main() {
 fn rust_version() -> Option<String> {
     let rustc = option_env!("RUSTC").unwrap_or("rustc");
     let output = Command::new(rustc).arg("-V").output().ok()?.stdout;
-    String::from_utf8(output).ok()
+    std::str::from_utf8(&output)
+        .ok()
+        .map(|o| o.split_whitespace().nth(1).unwrap().to_owned())
```
Comment (slowli) [inline exonum/build.rs]: Are you sure that `unwrap` is safe in this context? I'd return `None` in this case. Also, `?` could be used to simplify the code:

```rust
let rustc_output = std::str::from_utf8(&output).ok()?;
let version = rustc_output.split_whitespace().nth(1)?;
Some(version.to_owned())
```
